sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryParallelReadQuerySource.java

```diff
@@ -0,0 +1,178 @@
+import com.google.api.services.bigquery.model.TableReference;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO.ReadSessionOptions;
+import org.apache.beam.sdk.io.gcp.bigquery.BigQueryServices.DatasetService;
+import org.apache.beam.sdk.io.gcp.bigquery.BigQueryServices.JobService;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.options.ValueProvider;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A {@link org.apache.beam.sdk.io.BoundedSource} to read from BigQuery query results using the
+ * BigQuery parallel read API.
+ */
+public class BigQueryParallelReadQuerySource<T> extends BigQueryParallelReadSourceBase<T> {
+
+  // The maximum number of retries to poll a BigQuery job.
+  private static final int JOB_POLL_MAX_RETRIES = Integer.MAX_VALUE;
+
+  private static final Logger LOG = LoggerFactory.getLogger(BigQueryParallelReadQuerySource.class);
+
+  public static <T> BigQueryParallelReadQuerySource<T> create(
+      BigQueryServices bqServices,
+      SerializableFunction<SchemaAndRowProto, T> parseFn,
+      Coder<T> coder,
+      ReadSessionOptions readSessionOptions,
+      String stepUuid,
+      ValueProvider<String> query,
+      Boolean flattenResults,
+      Boolean useLegacySql) {
+    return new BigQueryParallelReadQuerySource<>(
+        bqServices,
+        parseFn,
+        coder,
+        readSessionOptions,
+        stepUuid,
+        query,
+        flattenResults,
+        useLegacySql);
+  }
+
+  private final String stepUuid;
+  private final ValueProvider<String> query;
+  private final Boolean flattenResults;
+  private final Boolean useLegacySql;
+
+  private transient AtomicReference<JobStatistics> dryRunJobStats;
+
+  private BigQueryParallelReadQuerySource(
+      BigQueryServices bqServices,
+      SerializableFunction<SchemaAndRowProto, T> parseFn,
+      Coder<T> coder,
+      ReadSessionOptions readSessionOptions,
+      String stepUuid,
+      ValueProvider<String> query,
+      Boolean flattenResults,
+      Boolean useLegacySql) {
+    super(bqServices, parseFn, coder, readSessionOptions);
+    this.stepUuid = checkNotNull(stepUuid, "stepUuid");
+    this.query = checkNotNull(query, "query");
+    this.flattenResults = checkNotNull(flattenResults, "flattenResults");
+    this.useLegacySql = checkNotNull(useLegacySql, "useLegacySql");
+    this.dryRunJobStats = new AtomicReference<>();
+  }
+
+  @Override
+  public long getEstimatedSizeBytes(PipelineOptions options)
+      throws InterruptedException, IOException {
+    BigQueryOptions bqOptions = options.as(BigQueryOptions.class);
+    return dryRunQueryIfNeeded(bqOptions).getQuery().getTotalBytesProcessed();
```

Comment:
Is the BytesProcessed the result table size?

Reply:
It's what the existing BigQueryQuerySource uses to estimate the table result size.

https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs

Reply:
Note that the existing Beam code reads from a deprecated version of this field. I've filed https://github.com/apache/beam/pull/5306 to update this.

Reply:
It is a bit strange since totalBytesProcessed is not the result table size, but it is not in the scope of this CL.

Reply:
There is no good way to know the size of the table without actually executing the query. This can be modeled a bit better when we move away from the Source model -- we can just produce the resulting table size as a tagged attribute of the PTransform. Until then, though...